superset/common/query_object.py

```diff
@@ -181,8 +181,10 @@ def __init__(
         self.order_desc = order_desc
         self.extras = extras
 
-        if config["SIP_15_ENABLED"] and "time_range_endpoints" not in self.extras:
-            self.extras["time_range_endpoints"] = get_time_range_endpoints(form_data={})
+        if config["SIP_15_ENABLED"]:
+            self.extras["time_range_endpoints"] = get_time_range_endpoints(
+                form_data=self.extras
+            )
```

Comment:
Oh wow, that would explain some of the problems people have been having with time range endpoints not working correctly. Passing `self.extras` as `form_data` feels slightly wrong; should we change the sig so that it's `def get_time_range_endpoints(extras: ExtraFormData)` or similar and perhaps add a `TypedDict` for it in `superset/utils/core.py` like here? https://github.com/apache/superset/blob/8ab45c90ccc04f51a0e7db12b622fb63a29ec27c/superset/utils/core.py#L175-L177

Reply:
@villebro the issue is that under other scenarios the full form data is required. We could have slots for both `form_data` and `extras` but I think that adds more confusion.

Reply:
I think there is a test case: https://github.com/apache/superset/blob/0a00153375fb69891c2a9f0115a33cdf5551b2d6/tests/query_context_tests.py#L170-L185

that captures the intention of the original code.

Do you mind adding a test case so your change is also tested?

Reply:
@ktmud this change captures the sentiment of the test, i.e., ensuring that the in `extras` the data is encoded as an enum rather than string, unless I'm confused regarding the `ChartDataQueryContextSchema` behavior.